pkg/app/web/src/components/application-detail-page/application-state-view/cloudrun-state-view/index.tsx

```diff
@@ -0,0 +1,189 @@
+import { Box, makeStyles } from "@material-ui/core";
+import clsx from "clsx";
+import dagre from "dagre";
+import { FC, useState } from "react";
+import { CloudRunResourceState } from "~/modules/applications-live-state";
+import { theme } from "~/theme";
+import { uniqueArray } from "~/utils/unique-array";
+import { CloudRunResource } from "./cloudrun-resource";
+import { CloudRunResourceDetail } from "./cloudrun-resource-detail";
+
+const useStyles = makeStyles((theme) => ({
+  root: {
+    display: "flex",
+    flex: 1,
+    justifyContent: "center",
+    overflow: "hidden",
+  },
+  stateViewWrapper: {
+    flex: 1,
+    display: "flex",
+    justifyContent: "center",
+    overflow: "hidden",
+  },
+  stateView: {
+    position: "relative",
+    overflow: "auto",
+  },
+  closeDetailButton: {
+    position: "absolute",
+    right: theme.spacing(1),
+    top: theme.spacing(1),
+    color: theme.palette.grey[500],
+  },
+}));
+
+export interface CloudRunStateViewProps {
+  resources: CloudRunResourceState.AsObject[];
+}
+
+const NODE_HEIGHT = 72;
+const NODE_WIDTH = 300;
+const STROKE_WIDTH = 2;
+const SVG_RENDER_PADDING = STROKE_WIDTH * 2;
+
+function useGraph(
+  resources: CloudRunResourceState.AsObject[],
+  showKinds: string[]
+): dagre.graphlib.Graph<{
+  resource: CloudRunResourceState.AsObject;
+}> {
+  const graph = new dagre.graphlib.Graph<{
+    resource: CloudRunResourceState.AsObject;
+  }>();
+  graph.setGraph({ rankdir: "LR", align: "UL" });
+  graph.setDefaultEdgeLabel(() => ({}));
+
+  const service = resources.find((r) => r.parentIdsList.length === 0);
+  const serviceID = service?.id || "";
+  resources.forEach((resource) => {
+    graph.setNode(resource.id, {
+      resource,
+      height: NODE_HEIGHT,
+      width: NODE_WIDTH,
+    });
+    if (resource.parentIdsList.length > 0) {
+      graph.setEdge(serviceID, resource.id);
+    }
+  });
```

Comment:
```suggestion
  const service = resources.find((r) => r.parentIdsList.length === 0);
  resources.forEach((resource) => {
    graph.setNode(resource.id, {
      resource,
      height: NODE_HEIGHT,
      width: NODE_WIDTH,
    });
    if (service && resource.parentIdsList.length > 0) {
      graph.setEdge(service.id, resource.id);
    }
  });
```
nits, sorry I just find out this one is safer 😄  it means we only make graph edge in case of service is not undefined, ignore `serviceID === ""` case.

Reply:
Exactly! Thank you!